src/index.js

```diff
@@ -136,17 +207,107 @@ module.exports = {
           (sum, userResult) => sum + userResult.data.betaVector[i], 0
         ) / userCount);
       }
+      
+      // calculate globalMeanY
+      const siteCount=userResults.length;  
+      var totalY=0;
+      var globalYCount=0;
+       
+      for (let i=0; i < siteCount; i += 1) {
+         totalY += userResults[i].data.localMeanY*userResults[i].data.localCount;
+         globalYCount += userResults[i].data.localCount;
+      } 
 
+      const globalMeanY = totalY/globalYCount;
+   
       /* eslint-disable no-console */
       console.log('Average beta vector:', averageBetaVector);
+      console.log('globalMeanY is :', globalMeanY);
       /* eslint-enable no-console */
+ 
+      // Extract X and y variable name list from userData
+      const y_label=opts.userResults[0].userData.__FEATURES__;
+      const X_label=Object.keys(opts.userResults[0].userData.files[0].tags);
+      console.log('X_label is', X_label);
+      console.log('y_label is',y_label);
 
       return {
         averageBetaVector,
+        globalMeanY,
+        X_label,
+        y_label,
+//        complete: true,
+      };
+    },
+    verbose: true,
+  }, {
+    // get sstNode from each local node and calculate the statistics
+    type: 'function',
+    fn(opts) {
+
+    const userResults = opts.userResults;
+    const X_label=opts.previousData.X_label;
+    const y_label=opts.previousData.y_label;
+
+    // get passed parameters from local nodes
+    const averageBetaVector = userResults[0].data.averageBetaVector;
```

Comment:
You’re iterating through `userResults` at least _10 times_ here. This would be easier to read, and more performant, to write as a single iteration:

```js
const totals = userResults.reduce(
  (
    memo,
    {
      data: {
        betaVector,
        localCount,
        pValue,
        pValueLocal,
        rSquared,
        rSquaredLocal,
        sseLocal,
        sstLocal,
        tValue,
        tValueLocal,
      },
    },
  ) => {
    memo.betaVector.push(betaVector);
    memo.globalYCount += localCount;
    memo.pValue.push(pValue);
    memo.pValueLocal.push(pValueLocal);
    memo.rSquared.push(rSquared);
    memo.rSquaredLocal.push(rSquaredLocal);
    memo.sseGlobal += sseLocal;
    memo.sstGlobal += sstLocal;
    memo.tValue.push(tValue);
    memo.tValueLocal.push(tValueLocal);

    return memo;
  },
  {
    betaVector: [],
    globalYCount: 0,
    pValue: [],
    pValueLocal: [],
    rSquared: [],
    rSquaredLocal: [],
    sseGlobal: 0,
    sstGlobal: 0,
    tValue: [],
    tValueLocal: [],
  }
);
```

This might be a bit verbose, but it’s clear what’s happening and cuts down on iteration. You can then easily use the values:

```js
const varError = (1 / totals.globalYCount - 2)) * totals.sseGlobal;
```

…and return everything at the end using `Object.assign`:

```js
return Object.assign(totals, {
  complete: true,
  pValueLocal,
  // ...
});
```